### PR TITLE
[ROCm] Enable context parallelism in pytorch TE

### PR DESCRIPTION
**Pull request by wangye805**
Also fixed a preview bug in workspace allocation for aotriton backend

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -2241,9 +2243,10 @@ def forward(
 
         #TODO: check whether context_parallel can be enabled with aotriton or CK backend
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/attention.py]: Update the TODO here. Say CK enabled and AOTriton not yet.

File: transformer_engine/common/fused_attn_rocm/fused_attn_aotriton.cpp [hunk 2]
```diff
@@ -630,6 +630,20 @@ void fused_attn_aotriton_bwd(
     workspace->data.dptr,
     &workspace_size,
     stream);
+
+  if (workspace_size > 0) {
```
Comment (wenchenvincent) [inline transformer_engine/common/fused_attn_rocm/fused_attn_aotriton.cpp:634]: Why do we add stuff with AOTriton in this PR?

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -2241,9 +2243,10 @@ def forward(
...
         if context_parallel:
-            assert (fused_attention_backend
-                == tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen
-                ), f"{fused_attention_backend} does not work with context parallelism!"
+            if not IS_HIP_EXTENSION:
+                assert (fused_attention_backend
+                    == tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen
+                    ), f"{fused_attention_backend} does not work with context parallelism!"
             assert (core_attention_bias_type == "no_bias"), \
```
Comment (ipanfilo) [inline transformer_engine/pytorch/attention.py:2249]: If CP is for CK only, should it be asserted here then?

File: transformer_engine/pytorch/attention.py (hunk 1 above)
Comment (wangye805) [reply transformer_engine/pytorch/attention.py]: Done.

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (wangye805) [reply transformer_engine/pytorch/attention.py:2249]: CP can be enabled for both CK and aotriton backend. Currently CP using aotriton will have some numerical issues. But we will try to fix this.

File: transformer_engine/common/fused_attn_rocm/fused_attn_aotriton.cpp (hunk 2 above)
Comment (wangye805) [reply transformer_engine/common/fused_attn_rocm/fused_attn_aotriton.cpp:634]: To fix a previous typo when I update backward workspace size extraction

**Review by wenchenvincent — APPROVED** [review]
LGTM.

**Pull request merged**